5.machine_learning/scripts/0.split_data.py

```diff
@@ -0,0 +1,237 @@
+random.seed(0)
+
+# Create new df which removes the holdout data from the plate_4_df which will be used to find random wells from rest of the data
+plate_4_df_filtered = plate_4_df[
+    ~(
+        (
+            (plate_4_df["Metadata_heart_number"] == random_heart_number)
+            & (plate_4_df["Metadata_cell_type"] == "Failing")
+        )
+        |
+        (
+            (plate_4_df["Metadata_heart_number"] == 7)
+            & (plate_4_df["Metadata_treatment"] == "DMSO")
+        )
+    )
+]
+
+# Generate random well per heart number to add to holdout_df
+random_wells = (
+    plate_4_df_filtered.groupby('Metadata_heart_number')['Metadata_Well']
+    .apply(lambda x: random.choice(sorted(x.unique())))  # Selecting a random well from sorted unique values
+    .reset_index(name='Random_Metadata_Well')
+)
+
+# Filter plate_4_df_filtered based on Metadata_heart_number and Metadata_Well in random_wells
+matched_rows = plate_4_df_filtered[
+    (plate_4_df_filtered['Metadata_heart_number'].isin(random_wells['Metadata_heart_number'])) &
+    (plate_4_df_filtered['Metadata_Well'].isin(random_wells['Random_Metadata_Well']))
+]
+
+# Prior to adding data into holdout_df to remove all holdout data at once, save random well data as "holdout2_data"
+matched_rows.to_csv(f"{output_dir}/holdout2_data.csv", index=False)
+
+# Add matching rows to the holdout data frame
+holdout_df = pd.concat([holdout_df, matched_rows], ignore_index=True)
+
+# Check shape and output
+print("There were", matched_rows.shape[0], "rows matching the random wells per heart")
+print("The shape of the holdout data frame after removing a random well per heart is", holdout_df.shape)
+holdout_df.head()
+
+
+# ## Remove all holdout data from the plate_4_df prior to splitting
+
+# In[7]:
+
+
+# Remove all rows from holdout data (using the data frame itself was not working)
+plate_4_df = plate_4_df[
+    ~(
+        (
+            (plate_4_df["Metadata_heart_number"] == random_heart_number)
+            & (plate_4_df["Metadata_cell_type"] == "Failing")
+        )
+        |
+        (
+            (plate_4_df["Metadata_heart_number"] == 7)
+            & (plate_4_df["Metadata_treatment"] == "DMSO")
+        )
+        |
+        (
+            (plate_4_df["Metadata_heart_number"].isin(random_wells["Metadata_heart_number"]))
+            & (plate_4_df["Metadata_Well"].isin(random_wells["Random_Metadata_Well"]))
+        )
+    )
+]
+
+print(plate_4_df.shape)
+plate_4_df.head()
+
+
+# ## Split remaining Plate 4 data into testing and training data
+
+# In[8]:
+
+
+# Set random state as 0 (same as the rest of the notebook)
+random_state = 0
+
+# Set the ratio of the test data to 30% (training data will be 70%)
+test_ratio = 0.30
+
+# Split the plate 4 data into training and test
+training_data, testing_data = train_test_split(
+    plate_4_df,
+    test_size=test_ratio,
+    stratify=plate_4_df[["Metadata_cell_type"]],
```

Comment:
That is a very good point. I think that based on the performance of the hold-out data we do not see that heart # as an influence to the prediction.

I will keep this in mind for the future though, since I didn't consider this before. I am curious to see if there is a struggle when it comes to apply to different plates with different heart #s.